### PR TITLE
Allow SPM to pass bearer token to Prometheus #4247

### LINKED ISSUE #4219: [Feature]: spm bearer token required for prometheus
### Requirement

I am using a global prometheus for spm where I need to present a bearer token to get access.
In kubernetes the token is available at /var/run/secrets/kubernetes.io/serviceaccount/token
Currently no parameter is available to define the token neither as value nor as file.

### Problem

spm gets a 401 when connected to prometheus

### Proposal

create a parameter like prometheus.bearer-token

### Open questions

_No response_

### PR DESCRIPTION
**Pull request by ChillOrb**
## Which problem is this PR solving?
- Resolves #4219  

## Short description of the changes
- Follow up to #4247

### CODE REVIEW

**Comment by ChillOrb** [comment]
@albertteoh I tried to revert submodule [change in /jaegar-ui ](https://github.com/jaegertracing/jaeger/pull/4247#pullrequestreview-1332340632) but it does not seem to me working. 
I tried to create[ new branch ](https://github.com/ChillOrb/jaeger/tree/dev2) and create a PR after adding changes but as soon as i make `git pull` i see submodule changes in `git status` . After reading [about submodules](https://git-scm.com/book/en/v2/Git-Tools-Submodules) 

`By default, the git pull command recursively fetches submodules changes, as we can see in the output of the first command above. However, it does not update the submodules. This is shown by the output of the git status command, which shows the submodule is “modified”, and has “new commits”. What’s more, the brackets showing the new commits point left (<), indicating that these commits are recorded in MainProject but are not present in the local DbConnector checkout. To finalize the update, you need to run git submodule update:`

How should i proceed with this? 

Maybe i can pull in all submodules update using `git submodule update --remote` ?

**Comment by ChillOrb** [comment]
> > as i make `git pull`
> 
> @ChillOrb, just for my understanding, why is there a need to do a `git pull`?
> 
> Technically, you should be able to create a new PR from your [previous PR](https://github.com/jaegertracing/jaeger/pull/4247) with:
> 
> ```
> # From your jaeger fork root directory on "main" branch.
> $ git checkout -b dev
> $ wget https://github.com/jaegertracing/jaeger/pull/4247.diff -O /tmp/4247.diff
> $ git apply /tmp/4247.diff
> $ git add .
> $ git commit -sm "My commit message"
> $ git push origin dev
> # Then create PR from dev branch.
> ```
> 
> Then, after making changes for an existing PR:
> 
> ```
> $ git add .
> $ git commit -sm "My change"
> $ git push origin dev
> ```
> 
>  > How should i proceed with this?
> 
> I would try running these commands from your project root then push the changes (more generally, I would run this if my submodules go "out of sync"):
> 
> ```
> $ git submodule deinit -f .
> $ git submodule update --init
> ```

@albertteoh , issue was with my main branch and therefore propagated to dev . I forced reset my main from head of jaegar:main , synced dev with main and selectively added files from dev2 to my dev. Sorry for all the confusion. Took me sometime to get this sorted.

Really appreciate you helping me out on this. 

Please review my PR again

File: plugin/metrics/prometheus/metricsstore/reader_test.go
Comment (albertteoh) [inline plugin/metrics/prometheus/metricsstore/reader_test.go:369]: Instead of `filepath.Split(tokenFilePath)`, could we use `filepath.Dir(tokenFilePath)` to just get the `dir` so that we can recursively create the directory structure for the test file, then replace `filepath.Join(dir, file)` with just `tokenFilePath` (`file` isn't really needed)?

File: plugin/metrics/prometheus/metricsstore/reader.go
Comment (albertteoh) [inline plugin/metrics/prometheus/metricsstore/reader.go:273]: We'll need to add a couple of tests to cover the error cases here and on L287.

Comment (ChillOrb) [reply plugin/metrics/prometheus/metricsstore/reader.go:273]: @albertteoh To cover these i was thinking of adding back table test case [here](https://github.com/ChillOrb/jaeger/blob/2c56f164f8a40f4a9d8d901753131454bc5abe73/plugin/metrics/prometheus/metricsstore/reader_test.go#L361) and [here](https://github.com/ChillOrb/jaeger/blob/2c56f164f8a40f4a9d8d901753131454bc5abe73/plugin/metrics/prometheus/metricsstore/reader_test.go#L505) 

I was thinking of using `wantErr bool` field and to mock an error maybe i can mock the behavior where user provides incorrect file path ( path does not exist error). This would however add some clumsy `if else` blocks where i check `if wantErr{ require.Err()}else{require.NoErr()}`.

Does this sound ok?

Comment (albertteoh) [reply plugin/metrics/prometheus/metricsstore/reader.go:273]: I think a better approach is to add a seperate simple test to cover the error case. This should cover both lines:
```
func TestGetRoundTripperTokenError(t *testing.T) {
	tokenFilePath := "this file does not exist"

	_, err := getHTTPRoundTripper(&config.Configuration{
		TokenFilePath: tokenFilePath,
	}, nil)

	require.Error(t, err)

	pathErr := &os.PathError{}
	assert.ErrorAs(t, err, &pathErr)
	assert.EqualError(t, pathErr, fmt.Sprintf("open %s: no such file or directory", tokenFilePath))
}
```

File: plugin/metrics/prometheus/metricsstore/reader_test.go
Comment (albertteoh) [inline plugin/metrics/prometheus/metricsstore/reader_test.go:364]: nit: I just realised that `logger` is only used if TLS is enabled, which is not the case for this test. So we can remove this and just pass `nil` into `getHTTPRoundTripper`, to simplify the test a little.

Comment (ChillOrb) [reply plugin/metrics/prometheus/metricsstore/reader_test.go]: Done

File: plugin/metrics/prometheus/metricsstore/reader.go
Comment (ChillOrb) [reply plugin/metrics/prometheus/metricsstore/reader.go:273]: @albertteoh thanks, i was overthinking this

File: plugin/metrics/prometheus/metricsstore/reader.go
Comment (albertteoh) [inline plugin/metrics/prometheus/metricsstore/reader.go:279]: For the benefit of anyone else looking over this PR, as mentioned earlier, I'll create a separate PR to address removing this line, and improve the `TestGetRoundTripperTLSConfig` test.

**Review by albertteoh — DISMISSED** [review]
This looks good to me, thanks, @ChillOrb! I'll give other maintainers a chance to look over this PR before merging.

File: plugin/metrics/prometheus/metricsstore/reader_test.go
Comment (yurishkuro) [inline plugin/metrics/prometheus/metricsstore/reader_test.go:376]: why not call this *after* the test server is created and its ServerURL is known?

File: plugin/metrics/prometheus/metricsstore/reader_test.go
Comment (ChillOrb) [reply plugin/metrics/prometheus/metricsstore/reader_test.go]: I agree , how does _[this](https://github.com/ChillOrb/jaeger/blob/714f86397dfdea11361e5cc066a0e1577481a8bf/plugin/metrics/prometheus/metricsstore/reader.go#L288)_ look?

File: plugin/metrics/prometheus/metricsstore/reader_test.go
Comment (ChillOrb) [reply plugin/metrics/prometheus/metricsstore/reader_test.go]: I agree , removed .. thanks

File: plugin/metrics/prometheus/metricsstore/reader_test.go
Comment (ChillOrb) [reply plugin/metrics/prometheus/metricsstore/reader_test.go]: Yep , since url is not required i have simplified and removed it

File: plugin/metrics/prometheus/metricsstore/reader_test.go
Comment (ChillOrb) [reply plugin/metrics/prometheus/metricsstore/reader_test.go]: I feel having a temp file is better than creating a directory for this test case . _[This](https://github.com/ChillOrb/jaeger/blob/d858ca2f51a3bdaaff76a61feab186a397f60048/plugin/metrics/prometheus/metricsstore/reader_test.go#L364)_ the change i have made to the function... Please share your views, thanks

File: plugin/metrics/prometheus/metricsstore/reader.go
Comment (ChillOrb) [reply plugin/metrics/prometheus/metricsstore/reader.go:271]: I was not aware of this case , thanks for pointing this out. As per my understanding if Jaeger is deployed on K8 cluster then token rotation is configurable(@hourly by default) and thus ought to be watched. Is that correct? 

Also, what if Jaeger is hosted on some other orchestration platform? 

Need some help deciding on next steps here..

@yurishkuro @albertteoh

Comment (yurishkuro) [reply plugin/metrics/prometheus/metricsstore/reader.go:271]: I would recommend just creating an issue for it to see if there's actually a demand for auto-reload of Prom token. Implementation-wise it should be using the fswatcher introduced in #4260

File: plugin/metrics/prometheus/metricsstore/reader.go
Comment (yurishkuro) [inline plugin/metrics/prometheus/metricsstore/reader.go:289]: unnecessary verbosity
```suggestion
		return "", fmt.Errorf("failed to get token from file: %w", err)
```

**Review by yurishkuro — APPROVED** [review]

File: plugin/metrics/prometheus/metricsstore/reader_test.go
Comment (yurishkuro) [inline plugin/metrics/prometheus/metricsstore/reader_test.go:373]: @ChillOrb I made a few tweaks
* simplify temp file creation
* make sure to call file.Close() after writing content to flush the write
* change ConnectTimeout to 1sec

Comment (yurishkuro) [reply plugin/metrics/prometheus/metricsstore/reader_test.go:373]: and use const for wantBearer

**Comment by yurishkuro** [comment]
Thanks!

**Pull request merged**
